docs/proposals/local-cluster.md

```diff
@@ -0,0 +1,422 @@
+cluster:
+  # set number of nodes to 6 for faster e2e
+  numNodes: 6
+  # use Weave Net
+  cniPlugin: weave
+  # use some addons besides the default (dashboard)
+  addons: dashboard,https://example.com/my-favorite-addon.yaml,/path/to/local/addon.yaml
+```
+
+As an additional configuration example, `gce.sh` script bundled with
+`kubeadm-local-cluster` can be used to set up a cluster on a GCE
+instance that's prepared using `docker-machine`.
+
+## Implementation notes
+
+### Base image
+
+The `kubeadm-local-cluster` tool by default uses a prebuilt image to
+start the cluster.  There's an image per each supported Kubernetes
+version and supported architecture (`amd64`, `arm`, `arm64`, `ppc64le`
+and `s390x`). There's also an image for each architecture that doesn't
+contain prebuilt Kubernetes binaries.  The images are based on
+`gcr.io/google-containers/debian-base-$(ARCH):0.1`
+
+It's possible to customize the behavior of the tool by creating a
+config file named `~/.kubeadm-local-cluster/config.yaml`.  It's
+possible to override default config path via `LOCAL_CLUSTER_CONF`
+environment variable.
+
+When `source.build*` options are enabled, `local-cluster`
+invokes appropriate `build/run.sh make ...` commands in Kubernetes
+source tree, preventing the build container from copying back the
+binaries using `KUBE_RUN_COPY_OUTPUT=n`. The control container then
+pulls the binaries directly from build container.
+
+### Networking
+
+The `kubeadm-local-cluster` tool utilizes named docker network
+`k8s-kubeadm-local-cluster`. This network is always created or replaced upon
+the cluster startup using address range specified via `DIND_SUBNET`
+configuration variable. `kubeadm-local-cluster` makes docker's internal DNS
+server available to all the nodes in the cluster, so nodes can be
+referred by their hostnames: `kube-master`, `kube-node-1`,
+`kube-node-2` and so on.
+
+### Cluster snapshot mechanism
+
+In order to reduce the amount of time needed to restart the cluster,
+the `kubeadm-local-cluster` tool takes the snapshot of the cluster
+after setting it up.  The snapshot is made by means of preserving each
+node's `/var/lib/docker` directory as a docker volume and saving each
+node container's filesystem in an image. The volumes and images
+created by snapshotting mechanism are removed during
+`./local-cluster clean`.
+
+### Future directions
+
+The following features may be added later:
+
+* automatic PV creation like it's done in `minikube`, so most Helm
+  charts from
+  [kubernetes/charts](https://github.com/kubernetes/charts) repository
+  work out of the box
+* enabling feature gates for Kubernetes components
+* automatic installation of `helm` and `tiller`
+* support using `delve` to debug k8s components.  Should be an option
+  for building k8s with `GOGCFLAGS="-N -l"` plus `delve` binary in the
+  images. Possible problem: as of Go 1.7, trying to build e.g.
+  `kubectl` with `GOGCFLAGS="-N -l"` may result in stack overflow
+  in Go compiler.
```

Comment:
@ivan4th could you please add what @philips suggested about simulating extra load with kubemark to this list...